src/module.c

```diff
@@ -3047,6 +3050,76 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
     return REDISMODULE_OK;
 }
 
+/* --------------------------------------------------------------------------
+ * Key API for the stream type.
+ * -------------------------------------------------------------------------- */
+
+/**
+ * Adds an entry to a stream. Like XADD without trimming.
+ *
+ * - `key`: The key where the stream is (or will be) stored
+ * - `id`: If the `AUTOID` flag is set, this is where the assigned ID is
+ *   returned. Can be NULL if `AUTOID` is set, if you don't care to receive the
+ *   ID. If `AUTOID` is not set, this is the requested ID.
+ * - `flags`: A bit field of
+ *   - `REDISMODULE_STREAM_AUTOID`: Assign a stream ID automatically, like `*`
+ *     in the XADD command.
+ *   - `REDISMODULE_STREAM_NOMKSTREAM`: Don't create the stream if it doesn't
+ *     exist.
+ * - `argv`: A pointer to an array of size `numfields * 2` containing the
+ *   fields and values.
+ * - `numfields`: The number of field-value pairs in `argv`.
+ *
+ * Returns `REDISMODULE_OK` if an entry has been added. Returns
+ * `REDISMODULE_ERR` if the key is not opened for writing, if the key has a type
+ * other than stream, if the key doesn't exist in case the flag `NOMKSTREAM` is
+ * set or when attempting to create an entry with an id which is not allowed.
+ */
+int RM_StreamAdd(RedisModuleKey *key, RedisModuleStreamID *id, int flags,
+                 RedisModuleString **argv, int64_t numfields) {
+    /* Validate args */
+    if (!(key->mode & REDISMODULE_WRITE))
+        return REDISMODULE_ERR;
+    if (key->value && key->value->type != OBJ_STREAM)
+        return REDISMODULE_ERR;
+    if (key->value == NULL && (flags & REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR;
+    if (!(flags & REDISMODULE_STREAM_AUTOID) &&
+        (id == NULL || (id->ms == 0 && id->seq == 0)))
+        return REDISMODULE_ERR;
+
+    /* Create key if necessery */
+    if (key->value == NULL)
+        moduleCreateEmptyKey(key, REDISMODULE_KEYTYPE_STREAM);
+
+    stream *s = key->value->ptr;
+
+    /* We don't assume streamID and RedisModuleStreamID are identical.
+       Should we? */
+    streamID added_id;
+    streamID use_id;
+    streamID *use_id_ptr = NULL;
+    if (!(flags & REDISMODULE_STREAM_AUTOID)) {
+        use_id.ms = id->ms;
+        use_id.seq = id->seq;
+        use_id_ptr = &use_id;
+    }
+    if (streamAppendItem(s, argv, numfields, &added_id, use_id_ptr) == C_ERR) {
+        /* Current last ID is greater than or equal to 'use_id' */
+        return REDISMODULE_ERR;
+    }
+
+    if (id != NULL) {
+        id->ms = added_id.ms;
+        id->seq = added_id.seq;
+    }
+
+    /* Notify blocked clients that there is new data on this stream. */
+    /* signalKeyAsReady(key->db, key->key, OBJ_STREAM); */
```

Comment:
i think your decisions of not doing signalKeyAsReady, signalKeyAsReady, etc implicitly are right.
module should do these manually when done amending the stream.
maybe it should be clearly stated in the documentation of the add API though.

Reply:
OK, good. However, `RedisModule_SignalKeyAsReady() calls the signalKeyAsReady() with OBJ_MODULE as the 3rd arg. I'm not sure it works or if needs to be OBJ_STREAM.

Reply:
good catch. that seems like a flaw in the API.
it should have taken a `RedisModuleKey` argument rather than key name.
not sure what we can do now, other than add a new API and deprecate the old one.
@guybe7 WDYT?

Reply:
`signalKeyAsReady()` is called by `dbAdd()` (db.c) so it's done implicitly by `RM_ListPush()` and `RM_ZsetAdd()` (via `moduleCreateEmptyKey()`). These types (BLPOP, etc.) can only block if the key doesn't exists, because popping the last element deletes the key.

`signalKeyAsReady()` is also done implicitly when adding to a non-existing stream, also via `moduleCreateEmptyKey()`.

Thus I think it should be done implicitly also when adding to an existing stream. Stream is the only type where readers can be blocked even if the key exists (e.g. using XREAD with id `$`).

Makes sense?

Reply:
Related: #8356

Reply:
I still don't think we should implicitly call signalKeyAsReady in RM_StreamAdd.
I think the module should do it explicitly when done adding.

for lists that may be ok to count on dbAdd (executed once per command), but streams you can add to an existing stream (many times inside one call), and also, the client can be blocked on a key that exists (unlike list that blocks on a key that's missing)

@guybe7 WDYT?

p.s. is the fact that RedisModule_SignalKeyAsReady still calls signalKeyAsReady() with OBJ_MODULE not causing any harm now? seems to me that it does and the only reason the test in #8356 passes is because it relies the automatic one in dbAdd.
i think we need to modify RedisModule_SignalKeyAsReady and signalKeyAsReady, so that when it is called it skips that new check.

Reply:
I think it's inconsistent if StreamAdd (and ListPush and ZsetAdd) to an empty key signals implicitly, but adding to an existing stream doesn't. If the worries are about performance, we can set a flag in the key struct and signal when the key is closed. How about that?

For comparison: Signal to WATCHed keys (`signalModifiedKey()`) is done implicitly when a key opened for writing is closed.

@guybe7 any ideas?

> p.s. is the fact that RedisModule_SignalKeyAsReady still calls signalKeyAsReady() with OBJ_MODULE not causing any harm now? seems to me that it does and the only reason the test in #8356 passes is because it relies the automatic one in dbAdd.

That's right. The test case relies on it being done implicitly, for the correct type.

Reply:
i suppose your suggestion about caching a flag in the key and doing it implicitly in close is ok.

can you also make a PR to solve the problem with RedisModule_SignalKeyAsReady?

Reply:
OK good. 😊 What's the problem in `RedisModule_SignalKeyAsReady()`? If all normal types are signaled implicitly, the only point of `RedisModule_SignalKeyAsReady()` is to signal for user-defined types...

Reply:
I've discussed it with Guy, at first we concluded that the code in `RedisModule_SignalKeyAsReady` is ok (no need to fix it).
since any modification on other types (list / streams) will always trigger an implicit SignalKeyAsReady, so the only reason to use that API is for a module that wants to release another client blocked on a module data type (which doesn't have an implicit SignalKeyAsReady).

But then we thought of another case, let's say there's a module that implements an LPOP, but one that's similarly to an XREAD can wait for new data even on a non-empty list (e.g. when a list has at least 5 elements).
This means that there's no implicit SignalKeyAsReady since a push to a non-empty list won't trigger it.
but on the other hand, if the only blocked client that's waiting for this case is a module, then calling RM_SignalKeyAsReady will release it (in that case, the argument to signalKeyAsReady is not about the type of the blocked key, but rather the identity of the entity that's blocked).

Maybe it would be nice to add a test like that a command like MODULE.LPOP 5 (that will unblock only when there are 5 elements to pop, and will block even if the list is not empty).

Reply:
`BLPOPN key 5` &ndash; It's so nice it deserves to be an official command :D

Jokes aside, that was a good discussion. Thanks for sharing. Adding a test case sounds like a good idea.

Reply:
Here's the test discussed: #8382. A module blocks until a list has 5 elements.